File: flatten-binary-tree.py
# Time Complexity: O(n)
# Space Complexity: O(n)
# Approach: Bottom up Recursion. At any arbitrary node, given that the left and right subtrees are flattened, use the tail node of the left subtree to connext to the right subtree and return the tail of the newly created linked list to the node. 
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def flatten(self, root):
        """
        :type root: TreeNode
        :rtype: None Do not return anything, modify root in-place instead.
        """
        # Edge
        if not root:
            return None
        
        def helper(node):
            if not node:
                return None
            
            left_tail, right_tail = None, None
            # Leaf
            if node.left is None and node.right is None:
                return node
            
            if node.left:
                left_tail = helper(node.left)
            if node.right:
                right_tail = helper(node.right)
                
            if left_tail:
                left_tail.right = node.right
                node.right = node.left
            node.left = None
                
            # Return
            if right_tail:
                return right_tail
            else:
                return left_tail
                
        helper(root)

File: assignments/src/FlattenBinaryTree.java
// Approach: Traverse all the left nodes then all the right nodes while maintaining nodes leftTail
// and rightTail. Re-wire the connections as -
// 1. leftmost leaf will have its child as the parent's right child
// 2. new formed leftmost's subtree is copied to its right sibling
// 3. leftmost leaf is set to null
// 4. return the rightTail
// Time: O(n)
// Space: O(n)
class FlattenBinaryTree {
    public void flatten(TreeNode root) {
        this.flattenTree(root);
    }

    private TreeNode flattenTree(TreeNode node) {
        if (node == null) return null;

        if (node.left == null && node.right == null) return node;

        TreeNode leftTail = this.flattenTree(node.left);
        TreeNode rightTail = this.flattenTree(node.right);

        if (leftTail != null) {
            leftTail.right = node.right;
            node.right = node.left;
            node.left = null;
        }

        return rightTail == null ? leftTail : rightTail;
    }
}